algorithms/src/polycommit/data_structures/powers.rs

```diff
@@ -0,0 +1,375 @@
+    marker::PhantomData,
+    path::PathBuf,
+    sync::Arc,
+};
+
+lazy_static::lazy_static! {
+    static ref DEFAULT_PATH: PathBuf = PathBuf::from(format!("{}/.aleo/powers_of_g", std::env::var("HOME").unwrap()));
+    static ref URLS: HashMap<usize, String> = {
+        let mut m = HashMap::new();
+        m.insert(1 << 16, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_16"));
+        m.insert(1 << 17, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_17"));
+        m.insert(1 << 18, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_18"));
+        m.insert(1 << 19, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_19"));
+        m.insert(1 << 20, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_20"));
+        m.insert(1 << 21, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_21"));
+        m.insert(1 << 22, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_22"));
+        m.insert(1 << 23, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_23"));
+        m.insert(1 << 24, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_24"));
+        m.insert(1 << 25, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_25"));
+        m.insert(1 << 26, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_26"));
+        m.insert(1 << 27, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_27"));
+        m.insert(1 << 28, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_28"));
+        m
+    };
+    static ref BASE_POWERS: &'static [u8] = include_bytes!("./powers_of_g_15");
+    static ref POWERS_TIMES_GAMMA_G: &'static [u8] = include_bytes!("./gamma_powers");
+}
+
+// Amount of powers contained in `POWERS_TIMES_GAMMA_G`.
+const NUM_POWERS_TIMES_GAMMA_G: usize = 84;
+
+/// An abstraction over a vector of powers of G, meant to reduce
+/// memory burden when handling universal setup parameters.
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""))]
+pub struct PowersOfG<E: PairingEngine> {
+    /// Filepath of the powers we're using.
+    file_path: String,
+    /// A handle to the file on disk containing the powers of G.
+    /// The handle is guarded to avoid read/write conflicts with potential
+    /// clones.
+    #[derivative(Debug = "ignore")]
+    file: Arc<RwLock<File>>,
+    /// Group elements of the form `{ \beta^i \gamma G }`, where `i` ranges from 0 to `degree`.
+    /// These are used for hiding.
+    pub powers_of_beta_times_gamma_g: Arc<RwLock<BTreeMap<usize, E::G1Affine>>>,
+    /// The degree up to which we have powers.
+    /// Mostly just used for ensuring we download in the proper order,
+    /// length checks are actually done by checking file metadata.
+    degree: Arc<RwLock<usize>>,
```

Comment:
Should we make a new struct and put that entire struct under a `Arc<RwLock<...>>`?